File: assignment 4/c/programs/9.c
#include <stdio.h>
#include <malloc.h>

struct Node
{
    int data;
    struct Node *next;
};

struct Node *createNode(int data)
{
    struct Node *newNode = (struct Node *)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

struct Node *insert(struct Node *head, int data)
{
    struct Node *newNode = createNode(data);

    if (head == NULL)
    {
        head = newNode;
    }
    else
    {
        struct Node *temp = head;
        while (temp->next != NULL)
        {
            temp = temp->next;
        }
        temp->next = newNode;
    }

    return head;
}

void printList(struct Node *head)
{
    while (head != NULL)
    {
        printf("%d ", head->data);
        head = head->next;
    }
    printf("\n");
}

int hasLoop(struct Node *head)
{
    struct Node *slow = head;
    struct Node *fast = head;

    while (fast != NULL && fast->next != NULL)
    {
        slow = slow->next;
        fast = fast->next->next;

        if (slow == fast)
        {
            return 1; // Loop detected
        }
    }

    return 0; // No loop
}

int main()
{
    struct Node *head = NULL;
    int num, data;

    printf("Enter the number of elements in the list: ");
    scanf("%d", &num);

    printf("Enter the elements:\n");
    for (int i = 0; i < num; i++)
    {
        scanf("%d", &data);
        head = insert(head, data);
    }

    // Creating a loop for demonstration purposes
    if (num >= 3)
    {
        struct Node *temp = head;
        for (int i = 0; i < num - 1; i++)
        {
            temp = temp->next;
        }
        temp->next = head;
    }

    if (hasLoop(head))
    {
        printf("The linked list contains a loop.\n");
    }
    else
    {
        printf("The linked list does not contain a loop.\n");
    }

    return 0;
}

File: assignment 4/python/programs/9.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


def create_node(data):
    return Node(data)


def insert(head, data):
    new_node = create_node(data)

    if head is None:
        head = new_node
    else:
        temp = head
        while temp.next is not None:
            temp = temp.next
        temp.next = new_node

    return head


def print_list(head):
    while head is not None:
        print(head.data, end=" ")
        head = head.next
    print()


def has_loop(head):
    slow = head
    fast = head

    while fast is not None and fast.next is not None:
        slow = slow.next
        fast = fast.next.next

        if slow == fast:
            return True  # Loop detected

    return False  # No loop


if __name__ == "__main__":
    head = None

    num = int(input("Enter the number of elements in the list: "))
    print("Enter the elements:")
    for _ in range(num):
        data = int(input())
        head = insert(head, data)

    # Creating a loop for demonstration purposes
    if num >= 3:
        temp = head
        for _ in range(num - 1):
            temp = temp.next
        temp.next = head

    if has_loop(head):
        print("The linked list contains a loop.")
    else:
        print("The linked list does not contain a loop.")
